src/Adapter/Product/Shop/CommandHandler/SetProductShopsHandler.php

```diff
@@ -0,0 +1,149 @@
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Shop\CommandHandler;
+
+use PrestaShop\PrestaShop\Adapter\Product\ProductDeleter;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Update\ProductShopUpdater;
+use PrestaShop\PrestaShop\Core\Domain\Product\Shop\Command\SetProductShopsCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Shop\CommandHandler\SetProductShopsHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Shop\Exception\ShopException;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+
+class SetProductShopsHandler implements SetProductShopsHandlerInterface
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var ProductDeleter
+     */
+    private $productDeleter;
+
+    /**
+     * @var ProductShopUpdater
+     */
+    private $productShopUpdater;
+
+    public function __construct(
+        ProductRepository $productRepository,
+        ProductDeleter $productDeleter,
+        ProductShopUpdater $productShopUpdater
+    ) {
+        $this->productRepository = $productRepository;
+        $this->productDeleter = $productDeleter;
+        $this->productShopUpdater = $productShopUpdater;
+    }
+
+    public function handle(SetProductShopsCommand $command): void
+    {
+        $productId = $command->getProductId();
+        $sourceShopId = $command->getSourceShopId();
+        $selectedShopIds = $command->getShopIds();
+        $initialShopIds = $this->productRepository->getAssociatedShopIds($productId);
+
+        $this->assertSourceShopIsAlreadyAssociated($sourceShopId, $initialShopIds);
+
+        $shopsToCopy = $this->findDifferentShopIds($selectedShopIds, $initialShopIds, $sourceShopId);
+        $shopsToRemove = $this->findDifferentShopIds($initialShopIds, $selectedShopIds, $sourceShopId);
+
+        // Remove non associated shops
+        $this->productDeleter->deleteFromShops($productId, $shopsToRemove);
+
+        // Copy data from source targets
+        foreach ($shopsToCopy as $targetShopId) {
+            $this->productShopUpdater->copyToShop(
+                $productId,
+                $sourceShopId,
+                $targetShopId
+            );
+        }
+    }
+
+    /**
+     * @param ShopId $sourceShopId
+     * @param ShopId[] $initialShopIds
+     */
+    private function assertSourceShopIsAlreadyAssociated(ShopId $sourceShopId, array $initialShopIds): void
+    {
+        if ($this->shopInArray($sourceShopId, $initialShopIds)) {
+            return;
+        }
+
+        throw new ShopException(sprintf(
+            'Source shopId must be one of current product shops. Could not find %d in the associated shops',
+            $sourceShopId->getValue()
+        ));
+    }
```

Comment:
You could move this in the command as well, and even use the same `InvalidProductShopAssociationException` (you can specify an error code if you really need to distinguish both use cases)

Reply:
I thought I already answered this.
This assertion requires having initialShopIds which are taken from database (and not from command input)